dm/dm/master/openapi_controller.go

```diff
@@ -0,0 +1,280 @@
+	worker := s.scheduler.GetWorkerBySource(sourceName)
+	if worker == nil {
+		return terror.ErrWorkerNoStart
+	}
+	taskNameList := s.scheduler.GetTaskNameListBySourceName(sourceName)
+	return s.scheduler.BatchOperateTaskOnWorker(ctx, worker, taskNameList, sourceName, pb.Stage_Running, true)
+}
+
+// nolint:unused
+func (s *Server) disableSource(ctx context.Context, sourceName string) error {
+	worker := s.scheduler.GetWorkerBySource(sourceName)
+	if worker == nil {
+		// no need to stop task if the source is not running
+		return nil
+	}
+	taskNameList := s.scheduler.GetTaskNameListBySourceName(sourceName)
+	return s.scheduler.BatchOperateTaskOnWorker(ctx, worker, taskNameList, sourceName, pb.Stage_Stopped, true)
+}
+
+func (s *Server) transferSource(ctx context.Context, sourceName, workerName string) error {
+	return s.scheduler.TransferSource(ctx, sourceName, workerName)
+}
+
+func (s *Server) checkTask(ctx context.Context, subtaskCfgList []*config.SubTaskConfig, errCnt, warnCnt int64) (string, error) {
+	return checker.CheckSyncConfigFunc(ctx, subtaskCfgList, errCnt, warnCnt)
+}
+
+// nolint:unparam,unused
+func (s *Server) createTask(ctx context.Context, subtaskCfgList []*config.SubTaskConfig) error {
+	return s.scheduler.AddSubTasks(false, pb.Stage_Stopped, subtaskCfgPointersToInstances(subtaskCfgList...)...)
+}
+
+// nolint:unused
+func (s *Server) updateTask(ctx context.Context, taskCfg *config.TaskConfig) error {
+	// TODO(ehco) no caller now , will implement later
+	return nil
+}
+
+// nolint:unparam,unused
+func (s *Server) deleteTask(ctx context.Context, taskName string) error {
+	sourceNameList := s.getTaskSourceNameList(taskName)
+	return s.scheduler.RemoveSubTasks(taskName, sourceNameList...)
+}
+
+// nolint:unused
+func (s *Server) getTask(ctx context.Context, taskName string) error {
+	// TODO(ehco) no caller now , will implement later
+	return nil
+}
+
+func (s *Server) getTaskStatus(ctx context.Context, taskName string, sourceNameList []string) ([]openapi.SubTaskStatus, error) {
+	workerStatusList := s.getStatusFromWorkers(ctx, sourceNameList, taskName, true)
+	subTaskStatusList := make([]openapi.SubTaskStatus, 0, len(workerStatusList))
+	for _, workerStatus := range workerStatusList {
+		if workerStatus == nil || workerStatus.SourceStatus == nil {
+			// this should not happen unless the rpc in the worker server has been modified
+			return nil, terror.ErrOpenAPICommonError.New("worker's query-status response is nil")
+		}
+		sourceStatus := workerStatus.SourceStatus
+		openapiSubTaskStatus := openapi.SubTaskStatus{
+			Name:       taskName,
+			SourceName: sourceStatus.GetSource(),
+			WorkerName: sourceStatus.GetWorker(),
+		}
+		if !workerStatus.Result {
+			openapiSubTaskStatus.ErrorMsg = &workerStatus.Msg
+			subTaskStatusList = append(subTaskStatusList, openapiSubTaskStatus)
+			continue
+		}
+		subTaskStatus := workerStatus.SubTaskStatus[0]
```

Comment:
will workerStatus.SubTaskStatus empty here?

Reply:
add check in https://github.com/pingcap/tiflow/pull/4510/commits/a6080c947cf8ec81309f44826f7e729e86a6ad42